pkg/kubelet/cri/remote/remote_image.go

```diff
@@ -80,71 +76,18 @@ func NewRemoteImageService(endpoint string, connectionTimeout time.Duration, tp
 		return nil, err
 	}
 
-	service := &remoteImageService{timeout: connectionTimeout}
-
-	if err := service.determineAPIVersion(ctx, conn, endpoint); err != nil {
-		return nil, err
-	}
-
-	return service, nil
-
-}
-
-// useV1API returns true if the v1 CRI API should be used instead of v1alpha2.
-func (r *remoteImageService) useV1API() bool {
-	return r.imageClientV1alpha2 == nil
-}
-
-// determineAPIVersion tries to connect to the remote image service by using the
-// highest available API version.
-//
-// A GRPC redial will always use the initially selected (or automatically
-// determined) CRI API version. If the redial was due to the container runtime
-// being upgraded, then the container runtime must also support the initially
-// selected version or the redial is expected to fail, which requires a restart
-// of kubelet.
-func (r *remoteImageService) determineAPIVersion(ctx context.Context, conn *grpc.ClientConn, endpoint string) error {
-	ctx, cancel := context.WithTimeout(ctx, r.timeout)
-	defer cancel()
-
-	klog.V(4).InfoS("Finding the CRI API image version")
-	r.imageClient = runtimeapi.NewImageServiceClient(conn)
-
-	if _, err := r.imageClient.ImageFsInfo(ctx, &runtimeapi.ImageFsInfoRequest{}); err == nil {
-		klog.V(2).InfoS("Using CRI v1 image API")
-
-	} else if status.Code(err) == codes.Unimplemented {
-		klog.V(2).InfoS("Falling back to CRI v1alpha2 image API (deprecated)")
-		r.imageClientV1alpha2 = runtimeapiV1alpha2.NewImageServiceClient(conn)
-
-	} else {
-		return fmt.Errorf("unable to determine image API version with %q, or API is not implemented: %w", endpoint, err)
-	}
-
-	return nil
+	return &remoteImageService{
+		timeout:     connectionTimeout,
+		imageClient: runtimeapi.NewImageServiceClient(conn),
```

Comment:
What do you think about keeping determineAPIVersion() and modifying it to report ImageFsInfo err codes.Unimplemented as not supported (and just remove the v1alpha part).  IMO there will still be a chance of old containerd's lying around that won't have v1 and we would benefit from the error log information. Additionally the info log out that we are using CRI v1 image API has some benefit. Noting we also have a possibility of image service and runtime service being split in the container runtime...

Reply:
Alright, changed the code as suggested :+1: